### DISCUSSION TITLE
react-compiler eslint rules causing an error

### DISCUSSION THREAD

**Discussion by maxwiseman** [pmndrs/zustand#2811] [Q&A] [answered]
I just updated my Next app to NextJS 15, and after enabling the React Compiler ESlint rules, I am getting an error on my Zustand store. This could just be me, but I think this is a fairly minimal setup, so I don't think I've done anything wrong. The error below is occurring on the underlined section of the last line. Everything still works fine, so I'm just `eslint-disable-next-line`ing it for now, but I'd like to figure out what's going on here. The specific ESlint rule causing this is `react-compiler/react-compiler`. I would make a bug report, but I can't seem reproduce it in a Stackblitz project. That might just be because I am setting up ESlint wrong in the Stackblitz though, I'm not sure. Any ideas?

```
eslint: Hooks must be the same function on every render, but this value may change over time to a different function. See https://react.dev/reference/rules/react-calls-components-and-hooks#dont-dynamically-use-hooks
```
```tsx
const useSettingsStore = create<{
    settings: UserSettings;
    setSettings: (arg0: UserSettings) => void;
  }>()(
    persist(
      (set) => ({
        settings: {
          colorScheme: "system",
        } as UserSettings,
        setSettings: (settings: UserSettings) => {
          set({ settings });
        },
      }),
      {
        name: "settings",
        storage: createJSONStorage(() => localStorage),
      },
    ),
  );

  const { settings, setSettings } = useSettingsStore();
                                    ^^^^^^^^^^^^^^^^
```

**Comment by dbritto-dev** [comment]
@maxwiseman would you mind sharing us a minimal repro on stackblitz?

**Reply by maxwiseman** [reply]
For some reason I can't get it to work on Stackblitz. I think because there doesn't seem to be a way to restart ESlint after it installs the react-compiler plugin? I did make a minimal reproduction in the form of a Github repo though, so if you'd like to clone that and run `npm i`, you should be able to see the issue in `app/use-settings.tsx`.
![2024-10-22 at 20 29 48@2x](https://github.com/user-attachments/assets/18fd96db-1b4f-4413-828f-54fcae1bfeaf)

**Comment by dai-shi** [comment, marked as answer]
The eslint warning is correct.

```ts
export function useSettings() {
  type UserSettings = {
    colorScheme: "system" | "light" | "dark";
  };
  const useSettingsStore = create<{
    settings: UserSettings;
    setSettings: (arg0: UserSettings) => void;
  }>()(
    persist(
      (set) => ({
        settings: {
          colorScheme: "system",
        } as UserSettings,
        setSettings: (settings: UserSettings) => {
          set({ settings });
        },
      }),
      {
        name: "settings",
        storage: createJSONStorage(() => localStorage),
      },
    ),
  );

  const { settings, setSettings } = useSettingsStore();
}
```

:point_up: You can't define a store inside a function.
It should be defined outside. 👇 

```ts
type UserSettings = {
  colorScheme: "system" | "light" | "dark";
};
const useSettingsStore = create<{
  settings: UserSettings;
  setSettings: (arg0: UserSettings) => void;
}>()(
  persist(
    (set) => ({
      settings: {
        colorScheme: "system",
      } as UserSettings,
      setSettings: (settings: UserSettings) => {
        set({ settings });
      },
    }),
    {
      name: "settings",
      storage: createJSONStorage(() => localStorage),
    },
  ),
);

export function useSettings() {
  const { settings, setSettings } = useSettingsStore();
}

**Reply by maxwiseman** [reply]
Ahh, yes, you're right. Don't know what I was thinking. Thanks!
